buildconfig/CMake/Sanitizers.cmake

```diff
@@ -9,54 +9,72 @@ set_property(CACHE USE_SANITIZER PROPERTY STRINGS
 string(TOLOWER "${USE_SANITIZER}" USE_SANITIZERS_LOWER)
 
 if(NOT ${USE_SANITIZERS_LOWER} MATCHES "off")
-    # Check we have a supported compiler
-    if(WIN32)
-        message(FATAL_ERROR "Windows does not support sanitizers")
-    endif()
 
     if(CMAKE_COMPILER_IS_GNUCXX AND (NOT CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 8))
         message(FATAL_ERROR "GCC 7 and below do not support sanitizers")
     endif()
 
-    # Check and warn if we are not in a mode without debug symbols
-    string(TOLOWER "${CMAKE_BUILD_TYPE}" build_type_lower)
-    if("${build_type_lower}" MATCHES "release" OR "${build_type_lower}" MATCHES "minsizerel" )
-        message(WARNING "You are running address sanitizers without debug information, try RelWithDebInfo")
+    if(MSVC)
+        if(MSVC_VERSION LESS 1927)
+            message(FATAL_ERROR "MSVC version below 16.7 doesn't support address sanitizer")
+        endif()
+    endif()
+
+    if(MSVC)
+        # some units don't compile using the /O2 flag
+        message("Replacing -O2 flag with -O1 so that code compiles with sanitizer")
+        string(REPLACE "/O2" "/O1" CMAKE_CXX_FLAGS_RELWITHDEBINFO ${CMAKE_CXX_FLAGS_RELWITHDEBINFO})
+        string(REPLACE "/O2" "/O1" CMAKE_C_FLAGS_RELWITHDEBINFO ${CMAKE_C_FLAGS_RELWITHDEBINFO})
+        set(CMAKE_CXX_FLAGS_RELWITHDEBINFO ${CMAKE_CXX_FLAGS_RELWITHDEBINFO} CACHE STRING "" FORCE)
+        set(CMAKE_C_FLAGS_RELWITHDEBINFO ${CMAKE_C_FLAGS_RELWITHDEBINFO} CACHE STRING "" FORCE)
 
-    elseif(${build_type_lower} MATCHES "relwithdebinfo")
+        if(${USE_SANITIZERS_LOWER} MATCHES "address")
+            add_compile_options($<$<OR:$<CONFIG:Debug>,$<CONFIG:RelWithDebInfo>>:/fsanitize=address>)
+            add_link_options("$<IF:$<CONFIG:Debug>,/wholearchive:clang_rt.asan_dbg_dynamic-x86_64.lib;/wholearchive:clang_rt.asan_dbg_dynamic_runtime_thunk-x86_64.lib;/wholearchive:vcasand.lib,/wholearchive:clang_rt.asan_dynamic-x86_64.lib;/wholearchive:clang_rt.asan_dynamic_runtime_thunk-x86_64.lib;/wholearchive:vcasan.lib>")
+            # Re additional switches enabled below on gcc:
+            # Equivalent of -fno-omit-frame-pointer in MSVC is /Oy- but documentation says it's not available for x64
+            # Don't think there is a VS equivalent of -fno-optimize-sibling-calls
+        else()
+            message(FATAL_ERROR "MSVC only supports address sanitizer")
+        endif()
+    else()
+        # Check and warn if we are not in a mode without debug symbols)
```

Comment:
Could we hoist this warning out of the if block, as it applies to both

Reply:
Sorry only just getting round to making these changes. I think this warning is only relevant to a single configuration generator such as ninja where you specify the build type? If you're using a multi configuration generator such as the Visual Studio generator then the configuration for all of the build types (release, debug etc) is created at the same time when you run cmake